File: sdks/ios/Sources/MatrixMediaSDK/MMClient.swift
import Foundation
import Combine

/// Entry point for the MatrixMedia SDK.
///
/// ## Usage
/// ```swift
/// let client = MMClient(
///     serverURL: URL(string: "https://mm.example.com")!,
///     tokenProvider: {
///         // Return a fresh Matrix OpenID token each time.
///         // NEVER cache the token -- this closure is called when a new token is needed.
///         return try await myMatrixClient.getOpenIDToken()
///     }
/// )
///
/// let user = try await client.authenticate()
/// let stream = try await client.joinStream(roomID: "!abc:example.org")
/// ```
///
/// ## Token Provider Contract
///
/// The `tokenProvider` closure is called whenever the SDK needs a fresh Matrix OpenID token.
/// This design avoids the stale-token bug from PTT v1 (Knowledge Base lesson #9):
/// **never cache the token; always return a fresh one from your Matrix SDK.**
///
/// ## Server URL
///
/// In release builds the server URL must use HTTPS. In debug builds, HTTP and localhost
/// are permitted for development.
@MainActor
public final class MMClient: ObservableObject {

    // MARK: - Public Properties

    /// Current connection state of the client.
    @Published public private(set) var connectionState: MMConnectionState = .disconnected

    /// The authenticated user, if any.
    @Published public private(set) var currentUser: MMUser?

    // MARK: - Internal Dependencies

    let serverURL: URL
    let tokenProvider: @Sendable () async throws -> MMOpenIDToken
    let apiClient: MMAPIClient
    let tokenManager: TokenManager

    // MARK: - Private State

    private var activeStreams: [String: MMStream] = [:]

    // MARK: - Init

    /// Creates a new MatrixMedia client.
    ///
    /// - Parameters:
    ///   - serverURL: Base URL of the mm-core server (e.g., `https://mm.example.com`).
    ///   - tokenProvider: Async closure that returns a **fresh** Matrix OpenID token.
    ///     Called on every authentication attempt. Must not cache tokens.
    public init(
        serverURL: URL,
        tokenProvider: @escaping @Sendable () async throws -> MMOpenIDToken
    ) {
        #if !DEBUG
        // PTT lesson #11: reject localhost in release builds.
        // Mobile devices cannot reach localhost -- must use LAN IP or public domain.
        precondition(
            serverURL.host != "localhost" && serverURL.host != "127.0.0.1",
            "MatrixMediaSDK: localhost is not reachable from device. Use a LAN IP or public domain."
        )
        precondition(
            serverURL.scheme == "https",
            "MatrixMediaSDK: HTTPS is required in release builds."
        )
        #endif

        self.serverURL = serverURL
        self.tokenProvider = tokenProvider
        self.tokenManager = TokenManager(serverURL: serverURL, tokenProvider: tokenProvider)
        self.apiClient = MMAPIClient(baseURL: serverURL, tokenManager: tokenManager)

        // Wire circular dependency: TokenManager needs MMAPIClient for token exchange.
        self.tokenManager.apiClient = self.apiClient
    }

    // MARK: - Authentication

    /// Authenticate with the mm-core server.
    ///
    /// Calls the `tokenProvider` to obtain a Matrix OpenID token, exchanges it with mm-core
    /// for an MM session JWT, and stores the session internally.
    ///
    /// - Returns: The authenticated ``MMUser``.
    /// - Throws: ``MMError/notAuthenticated`` if the token provider fails,
    ///   or ``MMError/serverError(code:message:)`` for server-side errors.
    @discardableResult
    public func authenticate() async throws -> MMUser {
        connectionState = .authenticating
        do {
            let user = try await tokenManager.authenticate()
            currentUser = user
            connectionState = .connected
            return user
        } catch {
            connectionState = .disconnected
            throw error
        }
    }

    // MARK: - Join Stream

    /// Join an existing stream as a listener.
    ///
    /// - Parameter roomID: The Matrix room ID containing the active stream.
    /// - Returns: An ``MMStream`` connected to the live audio/video feed.
    /// - Throws: ``MMError/notAuthenticated`` if not authenticated,
    ///   ``MMError/streamNotFound`` if no active stream in the room.
    public func joinStream(roomID: String) async throws -> MMStream {
        guard connectionState == .connected else {
            throw MMError.notAuthenticated
        }

        // 1. Get active stream info for the room
        let streamInfo = try await apiClient.getActiveStream(roomID: roomID)

        // 2. Join the stream (returns SFU token)
        let joinResponse = try await apiClient.joinStream(streamID: streamInfo.id)

        // 3. Create LiveKit bridge in subscriber mode (with E2EE info if present)
        // When mm-switch URL is present, the viewer can connect directly to
        // mm-switch via WebRTC for lower latency and server-side ad insertion.
        // NOTE: mm-switch direct WebRTC is not yet implemented in the iOS SDK.
        // The bridge falls back to LiveKit SFU for now. See: services/mm-switch/
        let bridge = LiveKitBridge(
            sfuURL: joinResponse.sfuURL,
            sfuToken: joinResponse.sfuToken,
            participantID: joinResponse.participantID,
            mode: .subscriber,
            e2ee: joinResponse.e2ee,
            switchURL: joinResponse.switchURL,
            switchSourceID: joinResponse.switchSourceID,
            switchViewerID: joinResponse.switchViewerID
        )

        // 4. Connect to the SFU (or mm-switch when implemented)
        try await bridge.connect()

        // 5. Create and track the stream
        let stream = MMStream(
            streamID: streamInfo.id,
            roomID: roomID,
            bridge: bridge,
            apiClient: apiClient,
            e2ee: joinResponse.e2ee
        )
        activeStreams[streamInfo.id] = stream
        return stream
    }

    // MARK: - Start Stream

    /// Create and start a new stream as a host.
    ///
    /// Requires microphone permission for audio streams. The SDK will check permission
    /// and throw ``MMError/microphonePermissionDenied`` if not granted.
    ///
    /// - Parameters:
    ///   - roomID: The Matrix room ID to stream in.
    ///   - config: Stream configuration (media type, title, max participants).
    /// - Returns: An ``MMStream`` connected as the host/publisher.
    /// - Throws: ``MMError/microphonePermissionDenied``, ``MMError/notAuthenticated``,
    ///   ``MMError/streamAlreadyActive``.
    public func startStream(
        roomID: String,
        config: MMStreamConfig = MMStreamConfig()
    ) async throws -> MMStream {
        guard connectionState == .connected else {
            throw MMError.notAuthenticated
        }

        // Check microphone permission for audio/video streams
        if config.mediaType == .audio || config.mediaType == .video {
            try await checkMicrophonePermission()
        }

        // 1. Create the stream on mm-core
        let createResponse = try await apiClient.createStream(roomID: roomID, config: config)

        // 2. Create LiveKit bridge in publisher mode (with E2EE info if present)
        // When mm-switch URL is present, the host can publish directly to
        // mm-switch via WebRTC (bypasses LiveKit for the streaming path).
        // NOTE: mm-switch direct publish is not yet implemented in the iOS SDK.
        // Falls back to LiveKit SFU for now. See: services/mm-switch/
        let bridge = LiveKitBridge(
            sfuURL: createResponse.sfuURL,
            sfuToken: createResponse.sfuToken,
            participantID: createResponse.participantID,
            mode: .publisher,
            e2ee: createResponse.e2ee,
            switchURL: createResponse.switchURL,
            switchSourceID: createResponse.switchSourceID
        )

        // 3. Connect and start publishing
        try await bridge.connect()
        try await bridge.enableMicrophone()

        // 4. Create and track the stream
        let stream = MMStream(
            streamID: createResponse.streamID,
            roomID: roomID,
            bridge: bridge,
            apiClient: apiClient,
            e2ee: createResponse.e2ee
        )
        activeStreams[createResponse.streamID] = stream
        return stream
    }

    // MARK: - Disconnect

    /// Disconnect from all active streams and clear the session.
    public func disconnect() async {
        for (_, stream) in activeStreams {
            await stream.leave()
        }
        activeStreams.removeAll()
        tokenManager.clearSession()
        currentUser = nil
        connectionState = .disconnected
    }

    // MARK: - Stream Queries

    /// Fetch active streams in a Matrix room.
    ///
    /// - Parameter roomID: The Matrix room ID.
    /// - Returns: Array of ``MMStreamInfo`` for active streams.
    public func listStreams(roomID: String) async throws -> [MMStreamInfo] {
        guard connectionState == .connected else {
            throw MMError.notAuthenticated
        }
        return try await apiClient.listStreams(roomID: roomID)
    }

    // MARK: - Recordings (VoD)

    /// List ready recordings in a Matrix room.
    ///
    /// - Parameters:
    ///   - roomID: The Matrix room ID.
    ///   - limit: Maximum number of recordings to return (default 20).
    /// - Returns: Array of ``MMRecording`` with status `.ready`.
    public func listRoomRecordings(roomID: String, limit: Int = 20) async throws -> [MMRecording] {
        guard connectionState == .connected else {
            throw MMError.notAuthenticated
        }
        let response = try await apiClient.listRoomRecordings(roomId: roomID, limit: limit)
        return response.recordings
    }

    /// Get details for a specific recording.
    ///
    /// - Parameter recordingID: The recording identifier.
    /// - Returns: The ``MMRecording``.
    public func getRecording(recordingID: String) async throws -> MMRecording {
        guard connectionState == .connected else {
            throw MMError.notAuthenticated
        }
        return try await apiClient.getRecording(recordingId: recordingID)
    }

    /// Delete a recording. Only the host of the original stream may delete its recording.
    ///
    /// - Parameter recordingID: The recording identifier.
    public func deleteRecording(recordingID: String) async throws {
        guard connectionState == .connected else {
            throw MMError.notAuthenticated
        }
        try await apiClient.deleteRecording(recordingId: recordingID)
    }

    // MARK: - Internal

    func removeStream(_ streamID: String) {
        activeStreams.removeValue(forKey: streamID)
    }

    // MARK: - Microphone Permission (PTT lesson #4)

    #if canImport(AVFoundation)
    private func checkMicrophonePermission() async throws {
        #if os(iOS) || os(macOS)
        @preconcurrency import AVFoundation
        let session = AVAudioSession.sharedInstance()
        switch session.recordPermission {
        case .granted:
            return
        case .denied:
            throw MMError.microphonePermissionDenied
        case .undetermined:
            let granted = await withCheckedContinuation { continuation in
                session.requestRecordPermission { granted in
                    continuation.resume(returning: granted)
                }
            }
            if !granted {
                throw MMError.microphonePermissionDenied
            }
        @unknown default:
            throw MMError.microphonePermissionDenied
        }
        #endif
    }
    #else
    private func checkMicrophonePermission() async throws {
        // No-op on platforms without AVFoundation
    }
    #endif
}

File: sdks/android/mm-sdk/src/main/kotlin/com/matrixmedia/sdk/MMClient.kt
package com.matrixmedia.sdk

import com.matrixmedia.sdk.internal.auth.TokenManager
import com.matrixmedia.sdk.internal.livekit.LiveKitBridge
import com.matrixmedia.sdk.internal.networking.MMApiClient
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Entry point for the MatrixMedia Android SDK.
 *
 * Manages authentication, stream discovery, and stream lifecycle.
 *
 * ## Usage
 *
 * ```kotlin
 * val client = MMClient(
 *     serverUrl = "https://mm.example.com",
 *     tokenProvider = { getMatrixOpenIdToken() }
 * )
 *
 * val user = client.authenticate(openIdToken)
 * val stream = client.joinStream(roomId = "!abc:example.org")
 * ```
 *
 * ## Token Provider
 *
 * The [tokenProvider] is called fresh for every API request that requires
 * authentication. This prevents the stale-token bug discovered in PTT v1
 * where cached tokens expired mid-session. The provider should return
 * the current valid Matrix OpenID token. The SDK handles MM session token
 * management (refresh at 80% TTL) internally.
 *
 * @param serverUrl Base URL of the MM server (e.g., `https://mm.example.com`).
 * @param tokenProvider Suspend function that returns a fresh [MMAuthToken].
 *   Called on every API request. Never caches tokens internally beyond the
 *   managed MM session token with auto-refresh.
 */
class MMClient(
    private val serverUrl: String,
    private val tokenProvider: suspend () -> MMAuthToken
) {
    private val _connectionState = MutableStateFlow(MMConnectionState.Disconnected)

    /** Current connection state of this client. */
    val connectionState: StateFlow<MMConnectionState> = _connectionState.asStateFlow()

    internal val tokenManager = TokenManager(serverUrl, tokenProvider)
    internal val apiClient = MMApiClient(serverUrl, tokenManager)

    init {
        validateServerUrl(serverUrl)
        // Wire the circular dependency: TokenManager needs MMApiClient for /auth/token
        // and /auth/refresh calls, while MMApiClient needs TokenManager for Bearer tokens.
        tokenManager.apiClient = apiClient
    }

    /**
     * Authenticate with the MM server using a Matrix OpenID token.
     *
     * Exchanges the OpenID token for an MM session JWT. The session token
     * is managed internally with auto-refresh at 80% TTL (15-minute tokens,
     * refreshed at 12 minutes).
     *
     * @param openIdToken Matrix OpenID token obtained from the homeserver.
     * @return The authenticated [MMUser].
     * @throws MMException.Network on connectivity failure.
     * @throws MMException.NotAuthorized if the homeserver rejects the token.
     */
    suspend fun authenticate(openIdToken: MMOpenIDToken): MMUser {
        _connectionState.value = MMConnectionState.Connecting
        return try {
            val user = tokenManager.authenticate(openIdToken)
            _connectionState.value = MMConnectionState.Connected
            user
        } catch (e: Exception) {
            _connectionState.value = MMConnectionState.Disconnected
            throw e
        }
    }

    /**
     * Join an active stream in the given Matrix room as a viewer.
     *
     * Discovers the active stream in the room, obtains an SFU token,
     * and connects to the LiveKit SFU as a subscriber.
     *
     * @param roomId Matrix room ID (e.g., `!abc:example.org`).
     * @return An [MMStream] representing the joined stream.
     * @throws MMException.StreamNotFound if no active stream exists in the room.
     * @throws MMException.StreamFull if the stream is at capacity.
     * @throws MMException.NotAuthenticated if [authenticate] was not called.
     * @throws MMException.SfuUnavailable if the SFU cannot be reached.
     */
    suspend fun joinStream(roomId: String): MMStream {
        tokenManager.ensureAuthenticated()

        val streams = apiClient.getRoomStreams(roomId)
        val active = streams.firstOrNull { it.status == "active" }
            ?: throw MMException.StreamNotFound()

        val joinResult = apiClient.joinStream(active.streamId)
        // When mm-switch URL is present, the viewer can connect directly to
        // mm-switch via WebRTC for lower latency and server-side ad insertion.
        // NOTE: mm-switch direct WebRTC is not yet implemented in the Android SDK.
        // Falls back to LiveKit SFU for now. See: services/mm-switch/
        val bridge = LiveKitBridge(
            sfuUrl = joinResult.sfuUrl,
            sfuToken = joinResult.sfuToken,
            isHost = false,
            e2ee = joinResult.e2ee,
            switchUrl = joinResult.switchUrl,
            switchSourceId = joinResult.switchSourceId,
            switchViewerId = joinResult.switchViewerId
        )
        bridge.connect()

        return MMStream(
            bridge = bridge,
            apiClient = apiClient,
            streamId = active.streamId,
            e2ee = joinResult.e2ee
        )
    }

    /**
     * Start a new stream in the given Matrix room as the host.
     *
     * Creates the stream on the MM server, obtains an SFU token with
     * publish permissions, connects to the LiveKit SFU, and enables
     * the microphone for audio streaming.
     *
     * The caller must ensure `RECORD_AUDIO` permission is granted before
     * calling this method. On Android 13+ (API 33), `POST_NOTIFICATIONS`
     * must also be granted for the foreground service notification.
     *
     * @param roomId Matrix room ID (e.g., `!abc:example.org`).
     * @param config Stream configuration. Defaults to audio-only.
     * @return An [MMStream] representing the started stream.
     * @throws MMException.MicrophonePermissionDenied if audio permission is missing.
     * @throws MMException.NotAuthenticated if [authenticate] was not called.
     * @throws MMException.SfuUnavailable if the SFU cannot be reached.
     */
    suspend fun startStream(
        roomId: String,
        config: MMStreamConfig = MMStreamConfig()
    ): MMStream {
        tokenManager.ensureAuthenticated()

        val result = apiClient.createStream(roomId, config, e2ee = config.e2ee)
        // When mm-switch URL is present, the host can publish directly to
        // mm-switch via WebRTC (bypasses LiveKit for the streaming path).
        // NOTE: mm-switch direct publish is not yet implemented in the Android SDK.
        // Falls back to LiveKit SFU for now. See: services/mm-switch/
        val bridge = LiveKitBridge(
            sfuUrl = result.sfuUrl,
            sfuToken = result.sfuToken,
            isHost = true,
            e2ee = result.e2ee,
            switchUrl = result.switchUrl,
            switchSourceId = result.switchSourceId
        )
        bridge.connect()
        bridge.enableMicrophone()

        return MMStream(
            bridge = bridge,
            apiClient = apiClient,
            streamId = result.streamId,
            e2ee = result.e2ee
        )
    }

    /**
     * List ready recordings in a Matrix room.
     *
     * Returns recordings with status [MMRecordingStatus.Ready], ordered newest-first.
     *
     * @param roomId Matrix room ID (e.g., `!abc:example.org`).
     * @param limit Maximum number of recordings to return. Defaults to 20.
     * @return List of [MMRecording] available for playback.
     * @throws MMException.NotAuthenticated if [authenticate] was not called.
     */
    suspend fun listRoomRecordings(roomId: String, limit: Int = 20): List<MMRecording> {
        tokenManager.ensureAuthenticated()
        return apiClient.getRoomRecordings(roomId, limit)
    }

    /**
     * Fetch details for a specific recording.
     *
     * @param recordingId The recording identifier.
     * @return The [MMRecording].
     * @throws MMException.StreamNotFound if the recording does not exist.
     * @throws MMException.NotAuthenticated if [authenticate] was not called.
     */
    suspend fun getRecording(recordingId: String): MMRecording {
        tokenManager.ensureAuthenticated()
        return apiClient.getRecording(recordingId)
    }

    /**
     * Delete a recording. Only the host of the original stream may delete its recording.
     *
     * @param recordingId The recording identifier.
     * @throws MMException.NotAuthorized if the caller did not host the original stream.
     * @throws MMException.StreamNotFound if the recording does not exist.
     * @throws MMException.NotAuthenticated if [authenticate] was not called.
     */
    suspend fun deleteRecording(recordingId: String) {
        tokenManager.ensureAuthenticated()
        apiClient.deleteRecording(recordingId)
    }

    // -- Creator profile (M1 Lightning) --

    /** Get the authenticated user's creator profile, or null if not onboarded. */
    suspend fun getCreatorProfile(): Map<String, Any?>? {
        tokenManager.ensureAuthenticated()
        return apiClient.getCreatorProfile()
    }

    /**
     * Publish (or clear with `null` / empty string) the authenticated creator's
     * LUD-16 Lightning Address. The server validates the format; donations route
     * directly via LNURL-pay when an address is set.
     */
    suspend fun updateLightningAddress(address: String?): Map<String, Any?> {
        tokenManager.ensureAuthenticated()
        return apiClient.updateCreatorProfile(address)
    }

    // -- Donations (M1 Lightning) --

    /**
     * Send a Lightning tip in sats. Returns the raw response. For the LNURL-pay
     * path, look in `result["invoice"]["bolt11"]` for the BOLT11 to display.
     */
    suspend fun sendLightningTip(streamId: String, amountSats: Int, message: String? = null): Map<String, Any?> {
        tokenManager.ensureAuthenticated()
        return apiClient.donateLightning(streamId, amountSats, message)
    }

    /**
     * Disconnect from the MM server and release all resources.
     *
     * After calling this, the client is no longer usable. Create a new
     * [MMClient] instance to reconnect.
     */
    suspend fun disconnect() {
        _connectionState.value = MMConnectionState.Disconnected
        tokenManager.clearSession()
    }

    private fun validateServerUrl(url: String) {
        if (url.isBlank()) {
            throw MMException.InvalidServerUrl(url)
        }
        val lower = url.lowercase()
        if (!lower.startsWith("http://") && !lower.startsWith("https://")) {
            throw MMException.InvalidServerUrl(url)
        }
    }
}
